### ISSUE TITLE
Scoped services do not resolve if appsettings.json and appsettings.Development.json exist

### ISSUE THREAD

**Issue by pillepalle1** [mayuki/Cocona#103]
# Steps to reproduce

## 1. Create a new console app

```sh
mkdir CoconaBug
cd CoconaBug

dotnet new console
dotnet add package Cocona

touch appsettings.json
touch appsettings.Development.json
```

## 2. Replace Program.cs 

```csharp
using Cocona;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;

var builder = Cocona.CoconaApp.CreateBuilder();

builder.Services.AddScoped<IScopedService, ScopedService>();

var app = builder.Build();

app.AddCommand("do-something", (
    IScopedService service) =>
{
    service.DoSomething();
});

app.Run();


internal interface IScopedService
{
    void DoSomething();
}

internal class ScopedService : IScopedService
{
    private readonly IConfiguration _config;

    public ScopedService(IConfiguration config)
    {
        _config = config;
    }
    
    public void DoSomething()
    {
        Console.WriteLine(_config["Test"]);
    }
}
```

## 3. Replace content of appsettings.json and appsettings.Development.json

appsettings.json

```json
{
  "Test": "TestTestTest.Plain"
}
```

appsettings.Development.json

```json
{
  "Test": "TestTestTest.Development"
}
```

## 4. Launch the program

```sh
DOTNET_ENVIRONMENT="Development" dotnet run do-something
```

Error message

```
Unhandled Exception: System.InvalidOperationException: Cannot resolve scoped service 'IScopedService' from root provider.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.ValidateResolution(Type serviceType, IServiceScope scope, IServiceScope rootScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.OnResolve(Type serviceType, IServiceScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Cocona.Command.Binder.CoconaParameterBinder.Bind(CommandDescriptor commandDescriptor, IReadOnlyList`1 commandOptionValues, IReadOnlyList`1 commandArgumentValues)
   at Cocona.Command.Dispatcher.Middlewares.CoconaCommandInvokeMiddleware.DispatchAsync(CommandDispatchContext ctx)
   at Cocona.Command.Dispatcher.Middlewares.HandleParameterBindExceptionMiddleware.DispatchAsync(CommandDispatchContext ctx)
   at Cocona.Command.Dispatcher.Middlewares.HandleExceptionAndExitMiddleware.DispatchAsync(CommandDispatchContext ctx)
```

**Comment by pillepalle1** [comment]
One might say this is not an issue, because each service could be injected as transient, but I am using helper methods such as

```csharp
    public static IServiceCollection AddApplication(this IServiceCollection services)
    {
        // Automagically add services via assembly scanning
        var executingAssembly = Assembly.GetExecutingAssembly();
        services.AddValidatorsFromAssembly(executingAssembly);
        services.AddMediatR(executingAssembly);
        
        // Manually add everything else
        services.AddScoped<IScopedService, ScopedService>();
        
        return services;
    }
```

to wire up DependencyInjection for AspNetCore. It would be great to be able to use them for CoconaApp as well.

**Comment by pillepalle1** [comment]
Turns out that every Environment but Development works. The reason is documented here:

https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.serviceprovideroptions.validatescopes?view=dotnet-plat-ext-7.0

and implemented here

https://github.com/dotnet/runtime/blob/6085dc2c1c1b678a6512565d351f926bfbdaf4a3/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L319

**Comment by timmkrause** [comment]
I do have the same scenario as you regarding reusability and don't want to change registration scopes.
On the other hand I also want to work with the Development environment.

Did you find a way to work around this?

```csharp
CoconaAppBuilder
    BootstrapHostBuilder.Apply()
        ConfigureDefaults()
            HostingHostBuilderExtensions.ConfigureDefaults() 
                .UseDefaultServiceProvider((context, options) =>
                {
                    bool isDevelopment = context.HostingEnvironment.IsDevelopment();
                    options.ValidateScopes = isDevelopment;
                    options.ValidateOnBuild = isDevelopment;
                });
```

This does not do the trick for me:

```csharp
builder.Services
    // Service registrations...
    .BuildServiceProvider(
        new ServiceProviderOptions
        {
            ValidateScopes = false, // Disable scope validation
            ValidateOnBuild = false // Disable build-time validation
        });
```

**Comment by nwoolls** [comment]
FWIW I got around this by registering my logic implementation with DI, then injecting an `IServiceProvider` into the command body, creating a scope from there, and then getting an instance of my service from the scope.

**Comment by Arnab-Developer** [comment]
When you add a service with `AddScoped` in an asp.net application, it creates a new object for every http request. But in a console application by default the scope is not defined. Scoped services are never resolved from the root provider if the environment is `Development`. This is the reason for which it is failing. You need to define that scope first and then you can get the object from the DI.

```csharp
app.AddCommand("do-something", (IServiceProvider serviceProvider) =>
{
    using var scope = serviceProvider.CreateScope();
    var service = scope.ServiceProvider.GetRequiredService<IScopedService>();
    service.DoSomething();
});
```

This problem is not related to `appsettings.json`.

**Comment by timmkrause** [comment]
@Arnab-Developer Thanks. Do you have background why the behavior is so different in "dev mode" and the root provider is not being used? I thought it was only a validation issue that could be bypassed in any way.
I am currently using constructor injection and have separated the command into a class as I like that approach better. Switching to ServiceProvider pattern just because of dev environment feels like a workaround. In that case I'd even consider "duplicating" registrations without scoped services.

**Comment by Arnab-Developer** [comment]
@timmkrause 

> Do you have background why the behavior is so different in "dev mode" and the root provider is not being used?

If we inject a scoped service in the `AddCommand` when the environment is `Production`, it will act like a singleton which will not be the correct behavior. So, even if we are not getting any exception in `Production` the application is not working correctly.

I think .NET do not throw exception in `Production` to avoid failure because real users are using production application. But when the environment is `Development` then it throws exception to notify the developers that one of your scoped service will behave like singleton if you inject this in `AddCommand`.

> Switching to ServiceProvider pattern just because of dev environment feels like a workaround.

You are correct, injecting `IServiceProvider` is a workaround. The recommended approach is injecting the `IServiceScopeFactory` into `AddCommand` which I found [here](https://learn.microsoft.com/en-us/dotnet/core/extensions/scoped-service) for `Development` and `Production` both environment.

```csharp
app.AddCommand("do-something", (IServiceScopeFactory serviceScopeFactory) =>
{
    using var scope = serviceScopeFactory.CreateScope();
    var service = scope.ServiceProvider.GetRequiredService<IScopedService>();
    service.DoSomething();
});
```

In summary:

A scoped service always needs a scope to behave properly. In case of ASP.NET the scope is automatically created with each http request. But if the application is a `console application` or a `background service` then there is nothing to create the scope automatically. In that case we should create that scope by injecting the `IServiceScopeFactory`.

If we do not create that scope in case of `console application` or a `background service` and the environment is `Development` then .NET will throw an exception. When the environment is `Production` there will be no exception, but the scoped service will behave like a singleton.
